File: code.py
import numpy as np
import matplotlib.pyplot as plt
from numba import jit
import pickle
from scipy.optimize import curve_fit as cf
import sys
#%matplotlib notebook
import time
 

kB=1
J=2
L=int(sys.argv[1])
T=float(sys.argv[2])
e=int(sys.argv[3])

Lx=L
Ly=2*L
N_sites=Lx*Ly
N_steps=10**e

start = time.time()

def nbr2D(L):
    N_sites=Lx*Ly
    nbrarr=np.zeros((N_sites,4),dtype=int)
    
    for i in range(N_sites):
        k1=i//Ly
        k2=i%Ly
        if 1<=k1<=Lx-2:
            nbrarr[i][2]=(k1-1)*Ly+k2 #up
            nbrarr[i][3]=(k1+1)*Ly+k2 #down
        else:
            a=(k1==0)
            nbrarr[i][2]=Ly*(Lx-2+a)+k2 #up
            nbrarr[i][3]=Ly*a+k2 #down
        if 1<=k2<=Ly-2:
            nbrarr[i][0]=i-1 #left
            nbrarr[i][1]=i+1 #right
        else:
            b=(k2==0)
            nbrarr[i][0]=(i-1+Ly*b)
            nbrarr[i][1]=(i+1-Ly*(1-b))
    return nbrarr
nbrarr=nbr2D(L)

@jit(nopython=True)
def ord_param(state_arr):
    m=0
    for j in range(Ly):
        rho_y=0
        for i in range(Lx):
            rho_y+=state_arr[i*Ly+j]/Lx
        m+=np.abs(rho_y-0.5)
    return 2*m/Ly


@jit(nopython=True)
def MC_update(state_arr,E,T): 
    beta=1/T

    i1=np.random.randint(N_sites) #site chosen
    
    if state_arr[i1]==1:
        i2=np.random.randint(4)
        if state_arr[nbrarr[i1][i2]]==0:
            delE=0
            for k in range(4):
                delE=delE+state_arr[nbrarr[i1][k]]
            for k in range(4):
                delE=delE-state_arr[nbrarr[nbrarr[i1][i2]][k]]
            delE=2*J*(delE+1)
            if (delE<=0 or np.random.random()<np.exp(-beta*delE)):
                E+=delE
                state_arr[i1]=0
                state_arr[nbrarr[i1][i2]]=1
    return E  

@jit(nopython=True)
def H(state_arr):
    eng=0
    for i in range(N_sites):
        for k in range(4):
            eng+=state_arr[i]*state_arr[nbrarr[i][k]]  
    return -J*eng  

@jit(nopython=True)
def get_data(T):
    #datafile=open('data.txt','a')
    
    pos_arr=np.arange(0,N_sites,1)
    np.random.shuffle(pos_arr)
    pos_arr=pos_arr[:N_sites//2]
    state_arr=np.zeros(N_sites)
    for i in range(N_sites//2):
        state_arr[pos_arr[i]]=1
    
    #state_arr=np.array((L*[1]+L*[0])*L)
    #np.savetxt(datafile,state_arr)

    Trlax=10000000
    
    E=H(state_arr)
    #E_mean=0
    #E2_mean=0
    #E4_mean=0
    m_mean=0
    m2_mean=0
    m4_mean=0
    
    for i in range(Trlax*N_sites):
        E=MC_update(state_arr,E,T)
    
    #return state_arr
    
    for i in range(N_steps):
        for j in range(N_sites):
            E=MC_update(state_arr,E,T)
        m=ord_param(state_arr)
        m_mean+=m
        m2_mean+=m*m
        m4_mean+=m*m*m*m
 
    return m_mean/(N_steps),m2_mean/(N_steps),m4_mean/(N_steps)
    

data_T=get_data(T)
end = time.time()


#plt.imshow(state_arr.reshape(Lx,Ly))
#plt.show()

#data_T=get_data(T)
np.savetxt(f'data_{T}.csv',data_T,delimiter=',')




print(f'runtime = {end-start} sec')

File: code.c
#include<stdio.h>
#include<math.h>
#include<stdlib.h>

int Lx,Ly,N_sites;
int J=1;
int state_arr[5000];
int nbrarr[5000][4];
float T;
int L;

void init_state()
{
    for(int i=0;i<N_sites;i++)
    {
        state_arr[i]=0;
    }
}

void init_ordered_state()
{
    for(int i=0;i<Lx;i++)
    {
        for(int j=0;j<Ly;j++)
        {
            state_arr[i*Ly+j]=1;
        }
    }
}
void init_randomize_state()
{
    srand(20);
    for(int i=0;i<N_sites/2;i++)
    {
        int pos=(int)((float)rand()/RAND_MAX*N_sites);
        //printf("%d ",pos);
        state_arr[pos]=1;
    }

}

void nbr2D(int nbrarr[5000][4])
{
    for(int i=0;i<N_sites;i++)
    {
        int k1=i/Ly;
        int k2=i%Ly;

        if(k1>=1 && k1<=Lx-2)
        {
            nbrarr[i][2]=i-Ly;
            nbrarr[i][3]=i+Ly;
        }
        else
        {
            int a=(k1==0);
            nbrarr[i][2]=Ly*(Lx-2+a)+k2;
            nbrarr[i][3]=Ly*a+k2;
        }
        if(k2>=1 && k2<=Ly-2)
        {
            nbrarr[i][0]=i-1;
            nbrarr[i][1]=i+1;
        }
        else
        {
            int b=(k2==0);
            nbrarr[i][0]=(i-1+Ly*b);
            nbrarr[i][1]=(i+1-Ly*(1-b));
        }
    }

}

float ord_param()
{
    float m=0;
    for(int j=0;j<Ly;j++)
    {
        float rho_y=0;
        for(int i=0;i<Lx;i++)
        {
            rho_y+=state_arr[i*Ly+j];
        }
        m+=fabs(rho_y/Lx - 0.5);
        //printf("%f\n",abs(-2));
    }
    return 2*m/Ly;
}

float H()
{
    float E=0;
    for(int i=0;i<N_sites;i++)
    {
        for(int j=0;j<4;j++)
            E+=state_arr[i]*state_arr[nbrarr[i][j]];
    }
    return -J*E;
}


float MC_update(float E)
{
    float beta=1/T;
    int i=(int)((float)rand()/RAND_MAX*N_sites);
    if(state_arr[i])
    {
        int j=(int)((float)rand()/RAND_MAX*4);
        if(!state_arr[nbrarr[i][j]])
        {
            float delE = 0;
            for(int k=0;k<4;k++)
                delE+= state_arr[nbrarr[i][k]] - state_arr[nbrarr[nbrarr[i][j]][k]];
            delE=2*J*(delE+1);

            if(delE<=0 || ((float)rand()/RAND_MAX)<exp(-beta*delE))
            {
                E+=delE;
                state_arr[i]=0;
                state_arr[nbrarr[i][j]]=1;
            }
        }
    }
    return E;
}



void get_data()
{
    int Trlax=10000000;
    float E=H();
    float m_mean=0;
    float m2_mean=0;
    float m4_mean=0;
    int N_steps=10000000;
    srand(99820);
    for(int i=0;i<Trlax;i++) //reach steady state
        for(int j=0;j<N_sites;j++)
            E=MC_update(E);
    printf("steady state reached, ord param =%f\n",ord_param());
    for(int i=0;i<N_steps;i++)
    {
        for(int j=0;j<N_sites;j++)
            E=MC_update(E);
        //printf("MC step %d\n",i);
        float m=ord_param();
        m_mean+=m;
        m2_mean+=m*m;
        m4_mean+=m*m*m*m;
    }
    float data_arr[]={m_mean/N_steps,m2_mean/N_steps,m4_mean/N_steps};
    printf("m=%f\nm2=%f\nm4=%f",data_arr[0],data_arr[1],data_arr[2]);
    //return data_arr;
}

int main()
{
    scanf("%d %f",&L,&T);
    Lx=L;Ly=2*L;N_sites=Lx*Ly;
    nbr2D(nbrarr);

    init_state();
    init_randomize_state();

    printf("ord param of initial state = %f\n",ord_param());
    get_data();
    //float* data_arr=get_data();

    //printf("m_mean = %f\nm2_mean = %f\nm4_mean = %f",data_arr[0],data_arr[1],data_arr[2]);
    return 0;
}


